src/cc/reno.rs

```diff
@@ -0,0 +1,176 @@
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+use std::time::Instant;
+
+use crate::cc;
+use crate::recovery::Sent;
+
+// Reno Congestion Control
+pub struct Reno<'a> {
+    name: &'a str,
+
+    trace_id: String,
+
+    congestion_window: usize,
+
+    bytes_in_flight: usize,
+
+    congestion_recovery_start_time: Option<Instant>,
+
+    ssthresh: usize,
+    /* TODO: ECN is not implemented.
+     * ecn_ce_counters: [usize; packet::EPOCH_COUNT] */
+}
+
+pub const CC_NAME_RENO: &str = "reno";
+
+impl cc::CongestionControlTrait for Reno<'_> {
+    fn new() -> Self
+    where
+        Self: Sized,
+    {
+        Reno {
+            name: CC_NAME_RENO,
+
+            trace_id: String::new(),
+
+            congestion_window: cc::INITIAL_WINDOW,
+
+            bytes_in_flight: 0,
+
+            congestion_recovery_start_time: None,
+
+            ssthresh: std::usize::MAX,
+            /* TODO: ECN is not implemented.
+             * ecn_ce_counters: [0; packet::EPOCH_COUNT], */
+        }
+    }
+
+    fn get_name(&self) -> &str {
+        self.name
+    }
+
+    fn set_trace_id(&mut self, id: String) {
+        self.trace_id = id;
+    }
+
+    fn cwnd(&self) -> usize {
+        self.congestion_window
+    }
+
+    fn set_cwnd(&mut self, cwnd: usize) {
+        self.congestion_window = cwnd;
+    }
+
+    fn ssthresh(&self) -> usize {
+        self.ssthresh
+    }
+
+    fn bytes_in_flight(&self) -> usize {
+        self.bytes_in_flight
+    }
+
+    fn set_bytes_in_flight(&mut self, bytes_in_flight: usize) {
+        self.bytes_in_flight = bytes_in_flight;
+    }
+
+    fn congestion_recovery_start_time(&self) -> Option<Instant> {
+        self.congestion_recovery_start_time
+    }
+
+    // set to minimum cwnd
+    fn set_min_cwnd(&mut self) {
+        self.congestion_window = cc::MINIMUM_WINDOW;
+    }
+
+    // OnPacketSentCC
+    fn on_packet_sent_cc(&mut self, bytes_sent: usize) {
+        self.bytes_in_flight += bytes_sent;
+
+        trace!(
+            "{} {} CC_DEBUG OnPacketSentCC() bytes_in_flight={}",
```

Comment:
All these logs shouldn't be specific to each CC implementation I think (again, implementing `Debug` would make this easier). It will also become harder to log this information when we add qlog support I think.

But also, do we really need them? Trace logs are already very expensive to the point where they become useless when doing performance measurements, so I'm not sure these would be useful.

Reply:
When we add a proper qlog support we can replace this with qlog functions but at this point this is particularly useful for debugging to check when the packet is sent bytes_in_flight is increased properly, so want to keep it.

Basically I need some log when each packets are sent/received. Yes trace!() usage is getting expensive now... If there is something can be used other than trace!() I'd love to try, or maybe I made another macro only for cc logging.

Reply:
`Debug` trait is implemented for `CongestionControl` and `trace_id` is removed (see related comment above for issues). But I still leave some trace!() logging for debuging cwnd and other CC variable changes.